### PR TITLE
JP-3011+JP-3113: IFU Extract1d updates

### PR DESCRIPTION
**Pull request by drlaw1558**
This PR resolves JP-3011 (source centroiding in IFU Extract1D) by implementing floating point centers and an optional source-finding algorithm using photutils DAOStarFinder.  It also resolves JP-3113 (MRS 1d residual fringe correction) by updating the residual fringe utilities code to handle 1d spectra and allowing the extract1d code to call this utility code when specified.  Both modifications are set to FALSE by default unless turned on by a user and/or a parameter reference file.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

Regression tests: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/

### CODE REVIEW

**Comment by jemorrison** [comment]
@drlaw1558 Over comments 
1. Has the extract 1d changes finding the centroids been tested on NIRSpec data ? 
2. Also I did not see a check in the code to run the 1d residual fringe correction only on MIRI data. You state in the comments it is only for MIRI data - but unless I missed it - you should have a check if the data is NIRSpec and then set ifu_rfcorr to False just to make sure a user with NIRSpec data does not run it by accident. 
3. Did you want to keep the plotting in the code. Most of the pipeline steps do not have plotting as an option, but I can see why it would be beneficial for off line processing to be able the plot the results. 
4. I scanned the new residual 1d code and I am assuming Patrick has tested it quite a lot. But I did not dig into the details of the code. I think we will found out if it works with experience and feed back from users.

File: jwst/residual_fringe/utils.py
Comment (nden) [inline jwst/residual_fringe/utils.py:1134]: `save_results` does not seem to be used anywhere or did I miss it?

File: jwst/residual_fringe/utils.py
Comment (nden) [inline jwst/residual_fringe/utils.py:4]: This adds matplotlib as a dependency. 
How important is it to have it in operations? 
If it's meant to be used by an end  user only function, it should be added as an optional dependency and wrapped  in an `if ...else` block. Also the plotting functionality in the functions should check if matplotlib is available and be skipped with a warning if not.

**Review by nden — CHANGES_REQUESTED** [review]

**Comment by nden** [comment]
What kind of tests should be added for this code?

**Comment by drlaw1558** [comment]
@jemorrison 
1) I tested it on one NIRSpec IFU cube from MAST (P330E).  There could be cases where it doesn't work, but in this test it found the source properly.
2) I just added such a check in ifu.py
3) Plots are just there for debugging.  It sounds like it may be simpler to just remove them.
4) I've done a fair bit of testing of this code, and indeed modified some of the interface from what Patrick originally sent.  It's working well in the tests I've done so far, with 1d spectra looking a lot less fringey and the 2d defringing unaffected (new routines were added to the utils.py, but no existing code changed).

File: jwst/extract_1d/ifu.py
Comment (drlaw1558) [reply jwst/extract_1d/ifu.py]: Yes, updated.

File: jwst/extract_1d/ifu.py
Comment (drlaw1558) [reply jwst/extract_1d/ifu.py]: Added

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py:989]: I've overhauled the docstrings on the new functions, I think they should now be in line with other code.

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: No idea- probably leftover from a previous version.

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Removed

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Removed

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: I've deleted all usage of plots and references thereto

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Removed

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Removed

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Removed

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Should now be reflected in docstring.

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: I don't think it is- deleted.

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py]: Removed dependency.

File: jwst/residual_fringe/utils.py
Comment (jemorrison) [inline jwst/residual_fringe/utils.py:1109]: you need a space before and after '=' or you will get flake8 issues

File: jwst/residual_fringe/utils.py
Comment (jemorrison) [inline jwst/residual_fringe/utils.py:1110]: same here

**Review by jemorrison — APPROVED** [review]
Just a few little flake8 issues to fix and it looks good. I am eager to try it out on MIRI data

File: jwst/extract_1d/extract_1d_step.py
Comment (drlaw1558) [reply jwst/extract_1d/extract_1d_step.py:152]: I'm a little hesitant here because I'm not sure about all the ways Extract1D can be calling models vs model containers, and if I add something it may cause an issue for some obscure spec mode I'm not familiar with.  Do you think the check on instr=MIRI before actually running the fringe code is enough?

File: jwst/residual_fringe/utils.py
Comment (drlaw1558) [reply jwst/residual_fringe/utils.py:850]: They could be, but I'd prefer to leave that to a future improvement if necessary.  All the infrastructure associated with new ref files is unnecessary at the moment since this is an optional non-default way of running the pipeline anyway.

First value is main fringe frequency, second is dichroic fringe frequency only relevant to ch 3+4.

Added some comments.

File: jwst/extract_1d/extract_1d_step.py
Comment (jemorrison) [reply jwst/extract_1d/extract_1d_step.py:152]: Yes I think adding the instr=MIRI is good enough

**Comment by nden** [comment]
@drlaw1558 LGTM now. I guess the CONSTANTS can use a few more words but those can be added. If you are not planning more changes I can resolve the conflicts and start a regtest run (once another run finishes).

**Comment by drlaw1558** [comment]
@nden No more changes planned, please go ahead.

**Review by nden — APPROVED** [review]
The regression tests pass. 
I'd wait for @hbushouse to take a look before merging.

**Comment by drlaw1558** [comment]
Additional testing revealed a few edge failure cases that could cause unnecessary crashes; the latest change embeds calls to 1d residual fringe correction in a try/except block so that any failures simply result in optional residual fringe correction not being applied.

File: jwst/residual_fringe/utils.py
Comment (hbushouse) [inline jwst/residual_fringe/utils.py:989]:
```suggestion
    res_fringes :  numpy array, required
        the 1D array with residual fringes
    weights : numpy array, required
        the 1D array of weights
    ffreq : float, required
        the central scan frequency
    dffreq :  float, required
        the one-sided interval of scan frequencies
    min_nfringes : int, required
        the minimum number of fringes to check
    max_nfringes : int, required
        the maximum number of fringes to check
    pgram_res : float, optional
        resolution of the periodogram scan in cm-1
    wavenum : numpy array, required
        the 1D array of wavenum

    Returns
    -------
    res_fringe_fit : numpy array
```

File: jwst/residual_fringe/utils.py
Comment (hbushouse) [inline jwst/residual_fringe/utils.py:1106]:
```suggestion
    flux : numpy array, required
        The 1D array of fluxes
    wavelength : numpy array, required
        The 1D array of wavelengths
    channel : integer, optional
        The MRS spectral channel
    dichroic_only : boolean, optional
        Fit only dichroic fringes
    max_amp : numpy array, optional
        The maximum amplitude array

    Returns
    -------
    output : numpy array
```

**Comment by hbushouse** [comment]
regtest run is clean, which is expected based on the fact that the new methods implemented in this PR are turned off by default via the new params that control them. Merging.

**Pull request merged**
